pkg/orchestrator/controller/controller_test.go

```diff
@@ -0,0 +1,407 @@
+			Expect(New(mockFederator).Start()).ToNot(Succeed())
+		})
+	})
+}
+
+func testClusterLifecycleNotifications() {
+	var (
+		controller   *controller
+		watchReactor *watchReactor
+	)
+
+	BeforeEach(func() {
+		controller = New(mocks.NewMockFederator(gomock.NewController(GinkgoT())))
+		watchReactor = newWatchReactor(controller)
+	})
+
+	AfterEach(func() {
+		watchReactor.close()
+		controller.Stop()
+	})
+
+	testOnAdd := func(clusterID string) {
+		controller.OnAdd(clusterID, &rest.Config{})
+
+		Expect(controller.clusterWatchers).Should(HaveKey(clusterID))
+		Eventually(watchReactor.endpointsWatchStarted).Should(Receive())
+		Eventually(watchReactor.clustersWatchStarted).Should(Receive())
+	}
+
+	testOnRemove := func(clusterID string) {
+		testOnAdd("east")
+		stopChan := controller.clusterWatchers["east"].stopChan
+		clusterWorkQueue := controller.clusterWatchers["east"].clusterWorkQueue
+		endpointWorkQueue := controller.clusterWatchers["east"].endpointWorkQueue
+
+		controller.OnRemove("east")
+
+		Expect(controller.clusterWatchers).ShouldNot(HaveKey("east"))
+		Expect(stopChan).To(BeClosed())
+		Eventually(clusterWorkQueue.ShuttingDown).Should(BeTrue())
+		Eventually(endpointWorkQueue.ShuttingDown).Should(BeTrue())
+	}
+
+	When("a cluster is added", func() {
+		It("should start watches for the Submariner Endpoint and Cluster resources", func() {
+			testOnAdd("east")
+		})
+	})
+
+	When("a cluster is removed", func() {
+		It("should remove and close the clusterWatch", func() {
+			testOnRemove("east")
+		})
+	})
+
+	When("a cluster is updated", func() {
+		It("should restart the Submariner Endpoint and Cluster watches", func() {
+			testOnAdd("east")
+			watchReactor.reset()
+			prevStopChan := controller.clusterWatchers["east"].stopChan
+			prevClusterWorkQueue := controller.clusterWatchers["east"].clusterWorkQueue
+			prevEndpointWorkQueue := controller.clusterWatchers["east"].endpointWorkQueue
+
+			controller.OnUpdate("east", &rest.Config{})
+
+			Expect(controller.clusterWatchers).Should(HaveKey("east"))
+			Eventually(watchReactor.endpointsWatchStarted).Should(Receive())
+			Eventually(watchReactor.clustersWatchStarted).Should(Receive())
+
+			Eventually(prevClusterWorkQueue.ShuttingDown).Should(BeTrue())
+			Eventually(prevEndpointWorkQueue.ShuttingDown).Should(BeTrue())
+			Expect(prevStopChan).To(BeClosed())
+		})
+	})
+
+	When("a cluster is re-added", func() {
+		It("should stop and restart the Submariner Endpoint and Cluster watches", func() {
+			testOnRemove("east")
+			watchReactor.reset()
+			testOnAdd("east")
+		})
+	})
+}
+
+func testResourceDistribution() {
+	const clusterID = "east"
+
+	var (
+		cluster          *submarinerv1.Cluster
+		endpoint         *submarinerv1.Endpoint
+		distributeCalled chan bool
+		mockCtrl         *gomock.Controller
+		mockFederator    *mocks.MockFederator
+		controller       *controller
+		fakeClientset    *fakeSubm.Clientset
+	)
+
+	BeforeEach(func() {
+		cluster = newCluster()
```

Comment:
I find it weird that you test clusters and endpoints in the same context, they could be each in it's own nested context.
This gets exacerbated when you add the delete and tests for that.

Reply:
I plan to split them into separate functions.